jax/interpreters/pxla.py

```diff
@@ -460,6 +460,9 @@ def __init__(self,
       sharding_spec = _pmap_sharding_spec(aval.shape[0], aval.shape[0],
                                           1, None, sharded_aval, True)
 
+    if isinstance(device_buffers[0], tuple):
```

Comment:
Given this check, does the type annotation on `device_buffers` need to be updated, to something like `device_buffers: List[Union[PyLocalBuffer, Tuple[PyLocalBuffer]]]` ?

This kind of polymorphism feels a little unsafe. Does the caller always know which case it's in? If so, we could refactor this so that `ShardedDeviceArray.__init__` doesn't change, but instead we just have a separate (non-polymorphic) function to handle this case. That would also make clear under what conditions this code path would be hit, since it's not clear right now.

Reply:
Agreed this is unsafe, and I initially avoided this kind of polymorphism. But the reason I went this route is that there is already code out in the wild that passes a list of `device_put()` operations into this constructor.

So the options we have are either to (1) upstream logic that checks the jax version and passes a list of buffers or list of tuples accordingly, or (2) accept both types of inputs within jax. The latter seems much more reasonable to me given current constraints.

We could perhaps also add a deprecation warning for the list-of-buffers case and make this polymorphism temporary.

Reply:
A third option would be to fix the callers in the wild (where "the wild" is google code). How about doing that? We don't need a deprecation warning; this is an internal API folks shouldn't rely on.

Reply:
Hm, I don't think it's the same as the first option because I'm not proposing any jax version checks. Instead I'm proposing just to make the breaking change, and if trax breaks we can fix them directly (or tell them they need to fix it themselves because it's an internal API). The overriding point here is that we must be willing to make breaking changes to internal APIs, otherwise revising JAX gets too hard. This is one such case.

Reply:
Ah, understood.

Reply:
I should add: in OSS libraries can always pin a jax version if they need to, or as you suggest switch on jax versions. They should own that tech debt rather than us centralizing it in JAX :)